src/main/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounder.java

```diff
@@ -509,32 +551,70 @@ public int register(NoGood noGood) {
-				int atomId = atomStore.putIfAbsent(substitutedAtom);
-
-				if (currentAssignment != null) {
-					ThriceTruth truth = currentAssignment.getTruth(atomId);
-					if (atomId > maxAtomIdBeforeGroundingNewNoGoods || truth == null || !truth.toBoolean()) {
-						// Atom currently does not hold, skip further grounding.
-						// TODO: investigate grounding heuristics for use here, i.e., ground anyways to avoid re-grounding in the future.
-						if (!disableInstanceRemoval) {
-							removeAfterObtainingNewNoGoods.add(substitutedAtom);
-							continue;
-						}
-					}
+				final TerminateOrTolerate terminateOrTolerate = storeAtomAndTerminateIfAtomDoesNotHold(substitutedAtom, currentAssignment, remainingTolerance);
+				if (terminateOrTolerate.decrementTolerance) {
+					remainingTolerance--;
+				}
+				if (terminateOrTolerate.terminate) {
+					continue;
+				}
+			}
+			bindingResult.add(advanceAndBindNextAtomInRule(groundingOrder, orderPosition, originalTolerance, remainingTolerance, unified, currentAssignment));
+		}
+
+		return bindingResult;
+	}
+
+	private TerminateOrTolerate storeAtomAndTerminateIfAtomDoesNotHold(final Atom substitute, final Assignment currentAssignment, final int remainingTolerance) {
+		int decrementedTolerance = remainingTolerance;
+		if (currentAssignment != null) { // if we are not in bootstrapping
+			final int atomId = atomStore.putIfAbsent(substitute);
+			ThriceTruth truth = currentAssignment.isAssigned(atomId) ? currentAssignment.getTruth(atomId) : null;
+
+			if (heuristicsConfiguration.isDisableInstanceRemoval()) {
+				final Instance instance = new Instance(substitute.getTerms());
+				boolean isInWorkingMemory = workingMemory.get(substitute, true).containsInstance(instance);
+				if (isInWorkingMemory) {
+					return new TerminateOrTolerate(false, false);
+				}
+				if (truth != null && !truth.toBoolean()) {
+					return new TerminateOrTolerate(true, false);
+				}
+				if (--decrementedTolerance < 0) {
+					// terminate if more positive atoms are unsatisfied as tolerated by the heuristic
+					return new TerminateOrTolerate(true, true);
+				}
+			} else {
+				if (truth == null || !truth.toBoolean()) {
+					// Atom currently does not hold
+					removeAfterObtainingNewNoGoods.add(substitute);
+				}
+				if (truth == null && --decrementedTolerance < 0) {
+					// terminate if more positive atoms are unsatisfied as tolerated by the heuristic
+					return new TerminateOrTolerate(true, true);
 				}
+				// terminate if positive body atom is assigned false
+				return new TerminateOrTolerate(truth != null && !truth.toBoolean(), decrementedTolerance < remainingTolerance);
 			}
-			List<Substitution> boundSubstitutions = bindNextAtomInRule(rule, groundingOrder, orderPosition + 1, unified, currentAssignment);
-			generatedSubstitutions.addAll(boundSubstitutions);
 		}
+		return new TerminateOrTolerate(false, decrementedTolerance < remainingTolerance);
+	}
+
+	private static class TerminateOrTolerate {
```

Comment:
The reason for it to have it is that `storeAtomAndTerminateIfAtomDoesNotHold` needs to return two values, `terminate` and `decrementTolerance`.
In my opinion, this solution is better than the one we had before 66c653c42da48f162131d827253077412b621065, where I abused an `AtomicInteger` to have call-by-reference.
Do you have a proposal that is even better?

Reply:
Well, since `TerminateOrTolerate` is used such that if `terminate`holds, the value of `decrementTolerance` is discarded, this class really just needs to represent three different values. The simplest (and fastest) alternative is to use an `int` as return value (such that, e.g. 0 means neither terminate, nor decrement, 1 means terminate, 2 means decrement). Given the fact that the method using this class (i.e., `createBindings`) is called really often, I'd seriously consider using an `int`. Maybe using int constants similarly to:

```
	private final static int RETURN_NEITHER = 0;
	private final static int RETURN_TERMINATE = 1;
	private final static int RETURN_DECREMENT = 2;
```

Another possibility is to use a simple enum with 3 constants. That way, at least no new objects need to be created every time `createBindings` is called.

Reply:
Excellent idea! 👍